packages/legend-application-query/src/stores/QueryEditorStore.ts

```diff
@@ -292,11 +308,220 @@ export class QueryExportState {
       this.editorStore.applicationStore.notificationService.notifyError(error);
     } finally {
       this.persistQueryState.reset();
-      this.editorStore.setExportState(undefined);
+      this.editorStore.setSaveAsState(undefined);
+    }
+  }
+}
+
+export class QuerySaveState {
+  editorStore: QueryEditorStore;
+  lambda: RawLambda;
+  queryName: string;
+  allowUpdate = false;
+  onQueryUpdate?: ((query: Query) => void) | undefined;
+  decorator?: ((query: Query) => void) | undefined;
+  queryBuilderState: QueryBuilderState;
+  persistSaveQueryState = ActionState.create();
+
+  constructor(
+    editorStore: QueryEditorStore,
+    queryBuilderState: QueryBuilderState,
+    lambda: RawLambda,
+    config: QueryPersistConfiguration,
+  ) {
+    makeObservable(this, {
+      queryName: observable,
+      allowPersist: computed,
+    });
+
+    this.editorStore = editorStore;
+    this.queryBuilderState = queryBuilderState;
+    this.lambda = lambda;
+    this.allowUpdate = config.allowUpdate ?? false;
+    this.queryName = config.defaultName ?? 'New Query';
+    this.decorator = config.decorator;
+    this.onQueryUpdate = config.onQueryUpdate;
+  }
+
+  get allowPersist(): boolean {
+    return (
+      !this.persistSaveQueryState.isInProgress &&
+      Boolean(this.queryBuilderState.mapping) &&
+      this.queryBuilderState.runtimeValue instanceof RuntimePointer
+    );
+  }
+
+  async persistSaveQuery(): Promise<void> {
```

Comment:
let's use `generator` function (i.e. `flow` form). Also rename this function to `saveQuery`